Wraps for open and write to file

M-ximus opened the pull request:
Fixes #1096

M-ximus committed:
add myself to the members list

M-ximus committed:
Western name order

M-ximus committed:
Merge branch 'master' of https://github.com/MIPT-ILab/mipt-mips

M-ximus committed:
Merge branch 'master' of https://github.com/MIPT-ILab/mipt-mips

M-ximus committed:
Merge branch 'master' of https://github.com/MIPT-ILab/mipt-mips

M-ximus committed:
Wraps for open and write to file checks

pavelkryukov commented:
Could you please provide more meaningful name for your pull request?

pavelkryukov commented on simulator/kernel/mars/t/unit_test.cpp:
```diff
@@ -236,6 +236,27 @@ TEST_CASE( "MARS: open non existing file")
+static void write_buff_to_file
+(System* sys, uint64 descr, uint64 buff_ptr, uint64 chars_to_write)
+{
+  sys.sim->write_cpu_register( v0, 15u); //write to file
```
You should use '->' instead of '.' here.

pavelkryukov commented on simulator/kernel/mars/t/unit_test.cpp:
```diff
@@ -236,6 +236,27 @@ TEST_CASE( "MARS: open non existing file")
 }
 
+static void write_buff_to_file
+(System* sys, uint64 descr, uint64 buff_ptr, uint64 chars_to_write)
```
Please follow our code guidelines regarding indentation here.

pavelkryukov commented on simulator/kernel/mars/t/unit_test.cpp:
```diff
@@ -236,6 +236,27 @@ TEST_CASE( "MARS: open non existing file")
+  CHECK( sys.mars_kernel->execute() == Trap::NO_TRAP);
+  uint64 descriptor = sys.sim->read_cpu_register( v0);
+  CHECK( descriptor >= 3); // May be print line from it was called?
+  return descriptor;
```
Can we have a separate test for file opening instead of making checks inside the helper functions?

```c++
TEST( "MARS: open file")
{
    auto trap = open_file( /* ... */);
    CHECK( trap == Trap::NO_TRAP);
    CHECK( get_descriptor() >= 3);
    // Should close as well?
}

TEST( "MARS: open file and write")
{
    open_file( /* ... */);
    auto descriptor = get_descriptor();
    auto trap = write_file( /* ... */ );
    CHECK( get_descriptor() >= 3);
    // Should close as well?
}

TEST_CASE( "MARS: open, write, read and close a file")
{
// ... 
```

pavelkryukov commented on simulator/kernel/mars/t/unit_test.cpp:
```diff
@@ -236,6 +236,27 @@ TEST_CASE( "MARS: open non existing file")
+  sys.sim->write_cpu_register( a1, flags);
+  CHECK( sys.mars_kernel->execute() == Trap::NO_TRAP);
+  uint64 descriptor = sys.sim->read_cpu_register( v0);
+  CHECK( descriptor >= 3); // May be print line from it was called?
```
Please either remove the comment or mark it as TODO or FIXME.

pavelkryukov commented on simulator/kernel/mars/t/unit_test.cpp:
```diff
@@ -236,6 +236,27 @@ TEST_CASE( "MARS: open non existing file")
+  sys.sim->write_cpu_register( a0, filename_ptr);
+  sys.sim->write_cpu_register( a1, flags);
+  CHECK( sys.mars_kernel->execute() == Trap::NO_TRAP);
+  uint64 descriptor = sys.sim->read_cpu_register( v0);
```
Please use `auto` here.

pavelkryukov commented on simulator/kernel/mars/t/unit_test.cpp:
```diff
@@ -236,6 +236,27 @@ TEST_CASE( "MARS: open non existing file")
+}
+
+static uint64 open_file(System* sys, uint64 filename_ptr, uint64 flags)
+{
```
Should these functions be members of System class?

M-ximus committed:
Merge branch 'master' of https://github.com/MIPT-ILab/mipt-mips into M-ximus

M-ximus committed:
Wrap for read from file and fix some bugs

pavelkryukov commented on simulator/kernel/mars/t/unit_test.cpp:
```diff
@@ -236,59 +236,69 @@ TEST_CASE( "MARS: open non existing file")
+    sys->sim->write_cpu_register( v0, 13u); // open file
+    sys->sim->write_cpu_register( a0, filename_ptr);
+    sys->sim->write_cpu_register( a1, flags);
+    //return sys->sim->read_cpu_register( v0); //FIXME: tests failed
```
The test fails since you only updated registers without performing actual system call.
You may either leave the function as is with better name (`prepare_to_open_file`) or put the call inside (`mars_kernel->execute`).

pavelkryukov commented on simulator/kernel/mars/t/unit_test.cpp:
```diff
@@ -236,59 +236,69 @@ TEST_CASE( "MARS: open non existing file")
-    sys.sim->write_cpu_register( a0, descriptor);
-    sys.sim->write_cpu_register( a1, 0x2000);
-    sys.sim->write_cpu_register( a2, 11);
+    write_buff_to_file(&sys, descriptor, 0x2000, 11); // write string to file
```
Good unit tests should have one Arrangement, one Act and one Assertion section, and the problem here is that Act and Assertion phases are mixed. In other words, we test that "open" was correct, than we test that "write" was succsessful, than we check the data read etc. As a result, if the test fails, we won't be able to say what stage failed without examining the unit test source code. So, we should either leave one section (checking that read data is what expected) or add additional tests to check what open/close/write do not produce exceptions.

M-ximus committed:
Merge branch 'master' of https://github.com/MIPT-ILab/mipt-mips into M-ximus

M-ximus committed:
Merge branch 'master' of https://github.com/MIPT-ILab/mipt-mips into M-ximus

M-ximus committed:
Separated file operation tests

Wraps for open, close, write and read that return a Trap. Tests for
opening and writing to file were separated from big I/O file test.

pavelkryukov commented on simulator/kernel/mars/t/unit_test.cpp:
```diff
@@ -236,65 +236,116 @@ TEST_CASE( "MARS: open non existing file")
+    CHECK( trap == Trap::NO_TRAP);
+    CHECK( descr >= 3);
+
+    trap = write_buff_to_file( &sys, descr, 0x2000, 11); // write string to file
```
The comment is redundant here since it follows method name

pavelkryukov commented on simulator/kernel/mars/t/unit_test.cpp:
```diff
@@ -236,65 +236,116 @@ TEST_CASE( "MARS: open non existing file")
+    CHECK( descr >= 3);
+
+    trap = write_buff_to_file( &sys, descr, 0x2000, 11); // write string to file
+    CHECK( trap == Trap::NO_TRAP); // FIXME: Can file be checked by read(unistd.h)
```
`unistd.h` is POSIX-specific. We may use C++ library instead.

`FIXME` should be followed by the name of the person who is assigned to fix it. Please put mine ('pkryukov'`) here.

pavelkryukov commented on simulator/kernel/mars/t/unit_test.cpp:
```diff
@@ -236,65 +236,116 @@ TEST_CASE( "MARS: open non existing file")
     auto descriptor = sys.sim->read_cpu_register( v0);
+    CHECK( trap == Trap::NO_TRAP);
     CHECK( descriptor >= 3);
+    trap = write_buff_to_file( &sys, descriptor, 0x2000, 11); // write string to file
```
The comment is redundant here since it follows method name

pavelkryukov commented on simulator/kernel/mars/t/unit_test.cpp:
```diff
@@ -236,65 +236,116 @@ TEST_CASE( "MARS: open non existing file")
     descriptor = sys.sim->read_cpu_register( v0);
+    CHECK( trap == Trap::NO_TRAP);
     CHECK( descriptor >= 3);
+    trap = write_buff_to_file( &sys, descriptor, 0x2000, 11); //write string to file
```
The comment is redundant here since it follows method name

M-ximus committed:
Delete useless comments
